modules/apps/segments/segments-context-vocabulary/src/main/java/com/liferay/segments/context/vocabulary/internal/configuration/persistence/listener/BaseConfigurationModelListener.java

```diff
@@ -0,0 +1,123 @@
+import com.liferay.portal.kernel.util.LocaleThreadLocal;
+import com.liferay.portal.kernel.util.ResourceBundleUtil;
+import com.liferay.portal.kernel.util.Validator;
+
+import java.util.Dictionary;
+import java.util.Locale;
+import java.util.Objects;
+import java.util.ResourceBundle;
+
+import org.osgi.service.cm.Configuration;
+import org.osgi.service.cm.ConfigurationAdmin;
+
+/**
+ * @author Janis Zhang
+ */
+public class BaseConfigurationModelListener
+	implements ConfigurationModelListener {
+
+	@Override
+	public void onBeforeSave(String pid, Dictionary<String, Object> properties)
+		throws ConfigurationModelListenerException {
+
+		String entityFieldName = String.valueOf(
+			properties.get("entityFieldName"));
+
+		if (Validator.isNull(entityFieldName)) {
+			throw new ConfigurationModelListenerException(
+				ResourceBundleUtil.getString(
+					_getResourceBundle(),
+					"please-enter-a-valid-session-property-name"),
+				_clazz, getClass(), properties);
+		}
+
+		if (_isDefined(
+				String.valueOf(properties.get("assetVocabularyName")),
+				String.valueOf(properties.get("companyId")), entityFieldName)) {
+
+			throw new DuplicatedSegmentsContextVocabularyConfigurationModelListenerException(
+				ResourceBundleUtil.getString(
+					_getResourceBundle(),
+					"this-field-is-already-linked-to-one-vocabulary"),
+				_clazz, getClass(), properties);
+		}
+	}
+
+	protected void init(ConfigurationAdmin configurationAdmin, Class<?> clazz) {
+		_configurationAdmin = configurationAdmin;
+		_clazz = clazz;
+	}
+
+	private ResourceBundle _getResourceBundle() {
+		Locale locale = LocaleThreadLocal.getThemeDisplayLocale();
+
+		return ResourceBundleUtil.getBundle(
+			"content.Language", locale, getClass());
+	}
+
+	private boolean _isDefined(
+			String assetVocabularyName, String companyId,
+			String entityFieldName)
+		throws ConfigurationModelListenerException {
+
+		try {
+			Configuration[] configurations =
+				_configurationAdmin.listConfigurations(
+					StringBundler.concat(
+						"(", ConfigurationAdmin.SERVICE_FACTORYPID, "=",
+						_clazz.getCanonicalName(), ")"));
+
+			if (configurations == null) {
+				return false;
+			}
+
+			for (Configuration configuration : configurations) {
+				Dictionary<String, Object> properties =
+					configuration.getProperties();
+
+				if ((Objects.equals(
+						assetVocabularyName,
+						properties.get("assetVocabularyName")) &&
+					 Objects.equals(
+						 entityFieldName, properties.get("entityFieldName"))) ||
+					(Objects.equals(
+						companyId,
+						String.valueOf(properties.get("companyId"))) &&
+					 Objects.equals(
+						 entityFieldName, properties.get("entityFieldName")))) {
+
+					return true;
+				}
+			}
+
+			return false;
+		}
+		catch (Exception exception) {
+			throw new ConfigurationModelListenerException(
+				exception.getMessage(), _clazz, getClass(), null);
+		}
+	}
+
+	private Class<?> _clazz;
+	private ConfigurationAdmin _configurationAdmin;
```

Comment:
@JanisZhang Please make `ConfigurationAdmin` field protected, and annotate it with `@Reference`.